### PR TITLE
cprnc should compare NaN

### LINKED ISSUE #4131: CPRNC does not handle NaN correctly
Currently, cprnc does not detect diffs involving NaN's. Not only it returns IDENTICAL for two vars containing both NaN (unlikely ok, but debatable), but it also returns IDENTICAL if one var has NaN and the other doesn't (definitely wrong). I'm attaching two mini files so you can run cprnc and verify it, but here's the files content and cprnc output:
```
$ ncdump src.nc 
netcdf src {
dimensions:
	ncol = 1 ;
variables:
	double x(ncol) ;
data:

 x = NaN ;
}
$ ncdump tgt.nc 
netcdf tgt {
dimensions:
	ncol = 1 ;
variables:
	double x(ncol) ;
data:

 x = 1 ;
}
$ ./cprnc -m src.nc tgt.nc 
 file 1=src.nc
 file 2=tgt.nc
 Dimension ncol matches
 x   (ncol)
                   1  (     1) (     1)
                   1                     NaN                     NaN
                   1   0.100000000000000E+01   0.100000000000000E+01
                   1  (     1) (     1)
          avg abs field values:                      NaN
                                   1.000000000000000E+00
************************************************************************************************************************************
  
SUMMARY of cprnc:
 A total number of      1 fields were compared
          of which      0 had non-zero differences
               and      0 had differences in fill patterns
               and      0 had different dimension sizes
 A total number of      0 fields could not be analyzed
 A total number of      0 fields on file 1 were not found on file 2.
 A total number of      0 fields on file 2 were not found on file 1.
  diff_test: the two files seem to be IDENTICAL 
```
I am positive this behavior is not desired (but correct me if I'm wrong), and I think we should change
```
diffcnt = count((vdiff>0) .and. mask(:,1))
```
to
```
! Note: (x .ne. x) seems silly, but it evals to true in case of NaN
diffcnt = count( ((vdiff>0) .or. (vdiff .ne. vdiff)) .and. mask(:,1))
```
in `compare_vars_mod.F90.in`.

[nc_files.zip](https://github.com/ESMCI/cime/files/7636000/nc_files.zip)

### PR DESCRIPTION
**Pull request by jedwards4b**
cprnc does not include NaN values in comparisons, a bug.

Test suite: hand tested using files provided in #4131  
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #4131 

User interface changes?: 

Update gh-pages html (Y/N)?:

### CODE REVIEW

**Comment by jedwards4b** [comment]
@bartgol Thank you for the bug report, I cannot add you as a reviewer but feel free to comment further.

**Comment by ekluzek** [comment]
I suggest you may want to make use of the shr_infnan_isnan function which is more robust at finding NaN's as sometimes compilers optimize out the (x /= x) trick at finding a NaN. So the shr_infnan_isnan function in share is a more robust way of doing that. However, if share code isn't included in the CPRNC build you probably can't do that.

**Review by fischer-ncar — APPROVED** [review]

**Comment by jedwards4b** [comment]
@ekluzek Based on your comment I have updated to use ieee_is_nan from the intrinsic ieee_arithmetic module.
I don't see any need to wrap this as is done in shr_infnan_mod since cprnc can always be built with gfortran.

**Comment by bartgol** [comment]
> I suggest you may want to make use of the shr_infnan_isnan function which is more robust at finding NaN's as sometimes compilers optimize out the (x /= x) trick at finding a NaN. So the shr_infnan_isnan function in share is a more robust way of doing that. However, if share code isn't included in the CPRNC build you probably can't do that.

That module isn't available to cprnc. But it's true that some compilers might optimize the check away (e.g., gfortran with `-ffast-math`, even with -O0). Perhaps the most reliable way of doing this is to set a control var to NaN, and then compare against that (but with FPE enabled, you can't just do `my_nan = 0; my_nan = my_nan/my_nan`, cause it would throw, so you need to init my_nan with the corresponding bit sequence for the given real type).

Imho, `x .ne. x` is the easiest solution (works for all scalar types, and fails only with specific opt flags).

**Comment by bartgol** [comment]
> @ekluzek Based on your comment I have updated to use ieee_is_nan from the intrinsic ieee_arithmetic module. I don't see any need to wrap this as is done in shr_infnan_mod since cprnc can always be built with gfortran.

Is that guaranteed to exist for all major compilers though? Some libs might build cprnc "on the fly", using the same compiler used for the rest of the code, which might not be gfortran.

**Comment by bartgol** [comment]
Ugh, `ieee_is_nan` is not more powerful than `x .ne. x` though:

```
program test
  use, intrinsic :: ieee_arithmetic, only: ieee_is_nan

  real :: a,b,c
  read (*,*) a,b 
  c = a/b

  if (ieee_is_nan(c)) then
    print *,c, " is a NaN!"
  else
    print *,c, " is not a NaN"
  endif
end program test
```
built with gfortran, with `-ffast-math`, entering 0 for both a and b:
```
$ ./program 
0
0
              NaN  is not a NaN
```
That's the same output obtained with `if(c .ne. c)`.

Anyhow, I don't think we need an almighty solution. Hopefully, people will build cprnc without `-ffast-math` (or similar). I'm ok with the current version.

**Review by ekluzek — APPROVED** [review]

**Comment by ekluzek** [comment]
The use of the ieee intrinsic makes it the most clear what is happening. And it should be the most robust. If it doesn't work correctly for all compilers and compiler optimization, that is a compiler specific bug. So I think this is a good solution.

**Comment by ekluzek** [comment]
One question I have though is what if both files have a NaN -- what does it do then? @bartgol did you test that case? Unfortunately I think that will be a common one for us that we need to allow.

**Comment by jedwards4b** [comment]
If both files have NaN:
```
 ./cprnc -m /glade/scratch/jedwards/piobug/src.nc /glade/scratch/jedwards/piobug/src.nc 
 file 1=/glade/scratch/jedwards/piobug/src.nc
 file 2=/glade/scratch/jedwards/piobug/src.nc
 Dimension ncol matches
 x   (ncol)
          1        1  (     1) (     1) (     1) (     1)
                   1                     NaN                     NaN     NaN                    NaN 0.0E+00                    NaN
                   1                     NaN                     NaN                            NaN                            NaN
                   1  (     1) (     1)
          avg abs field values:                      NaN    rms diff:     NaN   avg rel diff(npos):  0.0E+00
                                                     NaN                        avg decimal digits(ndif):  0.0 worst:  Inf
 RMS x                                       NaN            NORMALIZED  0.0000E+00

************************************************************************************************************************************
  
SUMMARY of cprnc:
 A total number of      1 fields were compared
          of which      1 had non-zero differences
               and      0 had differences in fill patterns
               and      0 had different dimension sizes
 A total number of      0 fields could not be analyzed
 A total number of      0 fields on file 1 were not found on file 2.
 A total number of      0 fields on file 2 were not found on file 1.
  diff_test: the two files seem to be DIFFERENT 
```

**Comment by ekluzek** [comment]
Awesome thanks for this @jedwards4b. I'd say it's good to go.

**Comment by ekluzek** [comment]
Oh wait. It worked but it called them as being different. So I think that's a problem.

**Review by billsacks — COMMENTED** [review]
Thanks a lot for fixing this @jedwards4b !

What will the behavior here be now if both files have NaN values in a particular location? (I can see arguments either way as to whether this should be considered a difference or not, so for now am just wondering what will happen.)

Should we add a test for the new behavior (at least for the case where one file has a NaN and the other doesn't) (in tools/cprnc/test_inputs)? It's pretty easy to add a new test of cprnc, though I'm not sure others are familiar with this (admittedly outdated & clunky – predating CIME) cprnc test infrastructure.

**Comment by billsacks** [comment]
Oh, somehow I missed all of the discussion in this PR before I commented. So ignore my question about both files having NaN.

**Comment by jedwards4b** [comment]
@billsacks Could you add a test using the files /glade/scratch/jedwards/piobug/tgt.nc /glade/scratch/jedwards/piobug/src.nc
@ekluzek The latest change only considers NaN values if they differ between the two files.

File: tools/cprnc/compare_vars_mod.F90.in [hunk 1]
```diff
@@ -358,12 +358,20 @@ contains
...
-
+          diffcnt = 0
 #if {ITYPE}==TYPEDOUBLE || {ITYPE}==TYPEREAL
-          diffcnt = count((vdiff>0 .or. ieee_is_nan(vdiff)) .and. mask(:,1))
-#else
-          diffcnt = count(vdiff>0 .and. mask(:,1))
+          ! Count the NaN values only if they differ between files
+          do i=1,s1
+             if(mask(i,1)) then
+                if(ieee_is_nan(buf(i,1)) .and. .not. ieee_is_nan(buf(i,2))) then
```
Comment (bartgol) [inline tools/cprnc/compare_vars_mod.F90.in:366]: Not really important, but one could just do
```
if (ieee_is_nan(buf(i,1) .ne. iee_is_nan(buf(i,2))) then
 diffcnt = diffcnt + 1
endif
```

**Comment by bartgol** [comment]
> What will the behavior here be now if both files have NaN values in a particular location? (I can see arguments either way as to whether this should be considered a difference or not, so for now am just wondering what will happen.)

Good point. I think the last mod (both NaN does not yield diff) is ok. The user can always pipe a call to something else (e.g. ncdump+grep) to spot any NaN. And I can see situations where you don't want to error out. E.g., you have a fixed size array, but depending on runtime params you only fill certain slices, leaving the rest unused (and the code could set arrays to NaN at init time, to enhance debugging).

**Comment by billsacks** [comment]
This looks good to me. I'll add a test for this, though it may not be until tomorrow.

**Comment by jedwards4b** [comment]
@billsacks awesome, thanks

**Comment by billsacks** [comment]
@jedwards4b I opened a PR into your branch: see https://github.com/jedwards4b/cime/pull/10

I added test files that have 3 variables, comparing:
- float variable with a NaN in one file and not in the other (reported as diff)
- double variable with a NaN in one file and not in the other (reported as diff)
- double variable with a NaN in both (reported as not differing)

I also ran all of the cprnc tests that are run via the run_tests script and compared against master. Other than the new test, all other tests are identical, as expected.

Once that PR is merged (assuming you approve), then I approve this PR.

**Pull request merged**